src/lib.rs

```diff
@@ -719,6 +719,37 @@ impl<T> Async<T> {
             optimistic(self.writable()).await?;
         }
     }
+
+
+    /// Waits until the I/O handle is readable.
+    ///
+    /// This function completes when a read operation on this I/O handle wouldn't block.
+    pub fn poll_readable(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let readable = self.source.readable();
+        pin!(readable);
```

Comment:
It is not good to create and destruct a new future each time poll is called. [The state of the future](https://github.com/stjepang/async-io/blob/5bde1c3d69304ac0dfa331439c5e9fccc9243e28/src/reactor.rs#L440-L442) is reset each time.

Reply:
I understand, but I don't see a different solution without changing the api to be `poll` based instead of `async`.

Reply:
Yeah, I thought it can do that by splitting [`Source::ready`](https://github.com/stjepang/async-io/blob/5bde1c3d69304ac0dfa331439c5e9fccc9243e28/src/reactor.rs#L439) into `fn poll_ready` that return `Poll` and `async fn ready` that call `poll_fn(|cx| self.poll_ready(cx, dir))`, but it seems difficult as `ready` takes `&self`.

So, I think the solution is probably not to provide `poll_{readable, writable}` on the async-io side, but to store the future returned by `{readable, writable}` on the user side.

Reply:
I can help with this tomorrow :)

Reply:
Fyi storing the future doesn't work, at least not without unsafe code as the future takes &self, you'd need to construct a self referential struct.